### PR TITLE
Use performance.now over Date.now in performance metrics

### LINKED ISSUE #16042: Tick events should use performance.now() over Date.now()
## What's the issue?

AMP runtime should use `performance.now()` instead of `Date.now()` when recording timestamps for [tick events.](https://github.com/ampproject/amphtml/blob/bca1ee26a23e2aa1ad8a5ee71fff0ab5fe09dfe8/extensions/amp-viewer-integration/TICKEVENTS.md) Clock changes such as those triggered by software like NTP, crossing daylight savings boundaries, crossing leap days / seconds, etc all can affect Date.now making it unreliable for precision timing.

I realize that the viewer's performance clock and the runtime's performance clock can be different. Maybe returning `performance.now() + performance.timing.navigationStart` would maintain the same interface but has the monotonic property that `Date.now()` lacks.

## What browsers are affected?

All browsers

## Which AMP version is affected?

All AMP versions afaik

### PR DESCRIPTION
**Pull request by alabiaga**
Fixes #16042 

- use performance.now in metrics
- fix missing jsdocs in performance-impl
- fix corresponding return types in url-replacements-impl due to jsdoc doc change
- update amp-access tests to polyfill performance.now as it is not present in a fakeWin
- update amp-access test to remove pubOrigin expect for 'http' as the pubOrigin for the iframe is now about:src since the fakeLocation is on win.testLocation as opposed to win.location when setting, as win.location on an iframe is not configurable. see describes.js#setup(env)
- remove opt_disableFallback param in amp-access.js#runAuthorization_() method as it is not used and remove related tests. This change was required as one of the test was failing, and not due to my changes as it fails on master.

### CODE REVIEW

File: extensions/amp-access/0.1/test/test-amp-access.js [hunk 1]
```diff
@@ -1474,7 +1473,7 @@ describes.fakeWin('AccessService analytics', {
 });
 
 
-describes.fakeWin('AccessService multiple sources', {
+describes.realWin('AccessService multiple sources', {
```
Comment (choumx) [inline extensions/amp-access/0.1/test/test-amp-access.js]: Why the changes in this file?

Comment (alabiaga) [reply extensions/amp-access/0.1/test/test-amp-access.js]: amp-access tests were failing because performance is now is not present in a fakeWin

**Comment by alabiaga** [comment]
PTAL. updated broken test in amp-access and polyfilled performance.now as opposed to using realWin. Thanks

File: extensions/amp-access/0.1/test/test-amp-access.js [hunk 2]
```diff
@@ -1493,7 +1470,9 @@ describes.fakeWin('AccessService multiple sources', {
     document = win.document;
     clock = sandbox.useFakeTimers();
     clock.tick(0);
-
+    // Polyfill performance.now() method as it is not available in fakeWin
+    // environment.
+    win.performance = {now: () => Date.now()};
```
Comment (choumx) [inline extensions/amp-access/0.1/test/test-amp-access.js]: You could also just add it to `fake-dom.js#FakeWindow`.

Comment (alabiaga) [reply extensions/amp-access/0.1/test/test-amp-access.js]: done. thanks!

File: extensions/amp-access/0.1/test/test-amp-access.js [hunk 3]
```diff
@@ -690,7 +659,9 @@ describes.fakeWin('AccessService pingback', {
     ampdoc = env.ampdoc;
     document = win.document;
     clock = sandbox.useFakeTimers();
-
+    // Polyfill performance.now() method as it is not available in fakeWin
+    // environment.
+    win.performance = {now: () => Date.now()};
```
Comment (kristoferbaxter) [inline extensions/amp-access/0.1/test/test-amp-access.js]: Since it ended up needing to be polyfilled in so many places, does it make sense to move this polyfill to `env.win` so each test doesn't need to repeat the polyfill?

Comment (alabiaga) [reply extensions/amp-access/0.1/test/test-amp-access.js]: Thanks for your feedback Kris. Will actually commented on that as well and I made the changes and moved this to fake-dom.js

File: extensions/amp-access/0.1/test/test-amp-access.js (hunk 2 above)
Comment (choumx) [reply extensions/amp-access/0.1/test/test-amp-access.js]: Looks like there are still a bunch of these in this file.

File: testing/fake-dom.js [hunk 4]
```diff
@@ -142,6 +142,10 @@ export class FakeWindow {
         }
       },
     });
+    // Polyfill performance.now() method as it would normally not be available
+    // in a non document environment. Used for tests using fakeWin that rely on
+    // the performance-impl service.
```
Comment (choumx) [inline testing/fake-dom.js]: This comment is slightly out of context and can be updated.

File: testing/fake-dom.js [hunk 5]
```diff
@@ -239,18 +246,10 @@ export class FakeWindow {
     this.requestAnimationFrame = raf;
   }
 
-  /**
-   * @param {string} type
-   * @param {function(!Event)} handler
-   * @param {(boolean|!Object)=} captureOrOpts
-   */
+  /** polyfill addEventListener. */
```
Comment (choumx) [inline testing/fake-dom.js]: Why replace the type declarations? Were they causing errors?

File: testing/fake-dom.js (hunk 4 above)
Comment (alabiaga) [reply testing/fake-dom.js]: what would you suggest?

File: extensions/amp-access/0.1/test/test-amp-access.js (hunk 2 above)
Comment (alabiaga) [reply extensions/amp-access/0.1/test/test-amp-access.js]: right..my editor cache needed to be cleared..for some reason it was showing my updated changes but this file wasn't actually edited. Fixed.

File: testing/fake-dom.js (hunk 5 above)
Comment (alabiaga) [reply testing/fake-dom.js]: yes lint errors

File: testing/fake-dom.js (hunk 4 above)
Comment (choumx) [reply testing/fake-dom.js]: How about just `// Polyfill performance.now with Date.now.`. I think most of the current comment is obvious since it's living in `FakeWindow`.

File: testing/fake-dom.js [hunk 6]
```diff
@@ -185,41 +189,44 @@ export class FakeWindow {
     /** @const */
     this.Date = window.Date;
 
+    this.setTimeout = function() {
```
Comment (choumx) [inline testing/fake-dom.js]: The JSDoc below doesn't seem useful when switching the order like this. We can either add the missing params (prefixed with `unused`) or remove the JSDoc outright. I think it's fine to remove since they're just standard window APIs.

Ditto below.

Comment (alabiaga) [reply testing/fake-dom.js]: done. Thanks

**Comment by mattpk** [comment]
@choumx @alabiaga @zhangsu @raywainman 

Thanks for addressing the bug so quickly. However, is it okay to just replace `Date.now()` with `performance.now()`? Performance.now returns a timestamp relative to `performance.timing.navigationStart`, which would be orders of magnitude different from `Date.now()`'s epoch time. Would this break any clients expecting the tick events to provide absolute timestamps?

A potential solution could be to return `performance.now() + performance.timing.navigationStart` so that the timestamp is still absolute. Alternatively, would it be possible to send the `performance.timing.navigationStart` value to the client so that we can know what the performance.now() timestamps are relative to? I can file an issue with a request if that sounds feasible.

File: src/service/performance-impl.js [hunk 7]
```diff
@@ -154,7 +154,7 @@ export class Performance {
       this.isMessagingReady_ = true;
 
       // Tick the "messaging ready" signal.
-      this.tickDelta('msr', this.win.Date.now() - this.initTime_);
+      this.tickDelta('msr', this.win.performance.now() - this.initTime_);
```
Comment (jridgewell) [inline src/service/performance-impl.js]: Oh man, this needs to be normalized to an int.
